File: framework/test2/validate/permute/permute.go
package main

import (
    . "missingjs.com/soda/leetcode"
    "missingjs.com/soda/unittest"
    "missingjs.com/soda/util"
)

var logger = util.Logger
var _used ListNode

var res []string

func permutation(chars []byte, n int) []string {
    res = make([]string, 0)
    buf := make([]byte, n)
    solve(chars, 0, buf, 0)
    return res
}

func solve(chars []byte, i int, buf []byte, j int) {
    if j == len(buf) {
        res = append(res, string(buf))
        return
    }
    for k := i; k < len(chars); k++ {
        chars[i], chars[k] = chars[k], chars[i]
        buf[j] = chars[i]
        solve(chars, i+1, buf, j+1)
        chars[i], chars[k] = chars[k], chars[i]
    }
}

func main() {
    // create tester by function
    work := unittest.CreateWork(permutation)
    // OR create by struct factory
    // work := unittest.CreateWorkForStruct(Constructor)
    work.SetValidator(unittest.Validators.ForSlice(nil, false))
    work.CompareSerial = true
    work.Run()
}

File: framework/test2/validate/permute/permute.ts
import {
    AvlTree,
    BinarySearchTree
} from '@datastructures-js/binary-search-tree';
import {
    PriorityQueue,
    MinPriorityQueue,
    MaxPriorityQueue
} from '@datastructures-js/priority-queue';
import { ListNode, NestedInteger, TreeNode } from 'soda/leetcode';
import { TestWork, Utils, Validators } from 'soda/unittest';

// step [1]: implement solution function
/**
 * @param {character[]} chars
 * @param {number} n
 * @return {character[]}
 */
function permutation(chars: string[], n: number): string[] {
    let res = [];
    let buf = Array(n).fill(null);
    solve(chars, 0, buf, 0, res);
    return res;
}

function solve(chars: string[], i: number, buf: string[], j: number, res: string[]) {
    if (j == buf.length) {
        res.push(buf.join(''));
        return;
    }
    for (let k = i; k < chars.length; ++k) {
        [chars[i], chars[k]] = [chars[k], chars[i]];
        buf[j] = chars[i];
        solve(chars, i+1, buf, j+1, res);
        [chars[i], chars[k]] = [chars[k], chars[i]];
    }
}

// step [2]: setup function/return/arguments
let taskFunc = permutation;
const work = TestWork.create<ReturnType<typeof taskFunc>>(taskFunc);
// let work = TestWork.forStruct(STRUCT);
work.validator = Validators.forArray<string>('string', false);
work.compareSerial = true;
console.log(work.run(Utils.fromStdin()));
